### ISSUE TITLE
Bad complexity of `zify`.

### ISSUE THREAD

**Issue by maximedenes** [coq/coq#8159]
`zify` blows up on small goals in `nat`, like:

```coq
Require Import Lia ZArith.

Goal 4 < 2 * 2 * 2 * 2 * 2 * 2 * 2 * 2 * 2 * 2 * 2 * 2 * 2.
zify.
```

I believe this bad complexity is due to the introduction of positivity hypotheses for each subterm. @fajb any thought about what we could do to improve it?

**Comment by maximedenes** [comment]
Maybe @ppedrot has an idea too.

**Comment by fajb** [comment]
An approach would be to detect redundant positivity constraints.
Suppose we already have positivity constraints C of the form (0 <= e) e.g (0 <= Z.of_nat t).
A constraint 0<= e is redundant if e is positive for the following rules:
1. e in C -> positive e
2. c is a positive constant -> positive(c)
3. positive e  -> positive (c * e) with c a positive constant
4. positive e1 /\ positive e2 -> positive (e1+e2)

Instead, you could also infer which subterm requires a positivity constraint.
This would mean constructing the minimal set C such that e is positive.

If I am right, that should not impact the decision power.

**Comment by maximedenes** [comment]
@fajb I see the idea (I think), but my Ltac skills are not up to the task. Do you think you could draft a patch? We can then refine it.

**Comment by fajb** [comment]
@maximedenes I doubt I have better Ltac skills. I'll give it a look.

**Comment by JasonGross** [comment]
I would guess you'd want something like the following?
```coq
Ltac assert_known_nonneg e :=
  lazymatch goal with
  | [ H : is_nonneg e |- _ ] => idtac
  | _
   => lazymatch e with
        | Zpos _ => idtac
        | Z0 => idtac
        | Zpos _ * ?x => assert_known_nonneg x
        | Z0 _ * ?x => assert_known_nonneg x
        | ?x * Zpos _ => assert_known_nonneg x
        | ?x * Z0 => assert_known_nonneg x
        | ?x + ?y => assert_known_nonneg x; assert_known_nonneg y
        end
  end.
```
and then you'd want to do something like `tryif assert_known_nonneg e then idtac else assert (is_nonneg e) by ...`

This assumes that you are doing a bottom-up sort of strategy, i.e., that you'll never be asked about `x + y` before having already seen `x` and `y`.

**Comment by pi8027** [comment]
This problem seems to have been addressed in #9856. At least, the problem is disappeared in @maximedenes's example. Should someone check other issues of `zify`?

**Comment by pi8027** [comment]
This problem seems to have been addressed in #9856. At least, the problem is disappeared in @maximedenes's example. Should someone check other issues of `zify`?

**Comment by pi8027** [comment]
If you still see the problem, please reopen this issue freely.
